Source/Core/VideoCommon/TextureDecoder_Common.cpp

```diff
@@ -434,55 +435,206 @@ static inline u32 DecodePixel_Paletted(u16 pixel, TLUTFormat tlutfmt)
+};
+#pragma pack(pop)
+
+static cmprTexelData TexDecoder_LoadCmprTexel(const u8* src, int s, int t, int widthBlks)
+{
+  u32 sBlk = s >> 3;
+  u32 tBlk = t >> 3;
+
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u32 blkS = s & 4;
+  u32 blkT = t & 4;
+  u32 blkOff = (blkT << 2) + (blkS << 1);
+
+  u32 offset = base + blkOff;
+
+  cmprTexelData ret;
+  std::memcpy(&ret, src + offset, sizeof(ret));
+  return ret;
+}
+
+static constexpr int CacheHashBits = 10;
+
+static constexpr u32 cmprCacheSet(u32 color1, u32 color2)
+{
+  return (color1 * color2 * 0x9E3779B1) >> (32 - CacheHashBits);
+}
+
+static constexpr u32 cmprCacheTag(u32 color1, u32 color2)
+{
+  return color1 | (color2 << 16);
+}
+
+struct CMPRCacheEntry
+{
+  u32 tag = {};
+  u32 colors[4] = {};
+};
+
+static CMPRCacheEntry cmprCache[1 << CacheHashBits] = {};
+
+static DOLPHIN_NO_INLINE void TexDecoder_DecodePixelCMPR_Miss(cmprTexelData data, u32 cacheSet,
+                                                              u32 cacheTag)
+{
+  u16 color1 = Common::swap16(data.color1);
+  u16 color2 = Common::swap16(data.color2);
+
+  int blue1 = Convert5To8(color1 & 0x1F);
+  int blue2 = Convert5To8(color2 & 0x1F);
+  int green1 = Convert6To8((color1 >> 5) & 0x3F);
+  int green2 = Convert6To8((color2 >> 5) & 0x3F);
+  int red1 = Convert5To8((color1 >> 11) & 0x1F);
+  int red2 = Convert5To8((color2 >> 11) & 0x1F);
+
+  auto& c = cmprCache[cacheSet].colors;
+  c[0] = MakeRGBA(red1, green1, blue1, 255);
+  c[1] = MakeRGBA(red2, green2, blue2, 255);
+  if (color1 > color2)
+  {
+    c[2] = MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1), DXTBlend(blue2, blue1), 255);
+    c[3] = MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2), DXTBlend(blue1, blue2), 255);
+  }
+  else
+  {
+    // color[3] is the same as color[2] (average of both colors), but transparent.
+    // This differs from DXT1 where color[3] is transparent black.
+    c[2] = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 255);
+    c[3] = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 0);
+  }
+
+  cmprCache[cacheSet].tag = cacheTag;
+}
+
+static u32* TexDecoder_CMPRCacheLookup(cmprTexelData data)
+{
+  // It's quite expensive to derive the four colors of each block.
+  // So we cache then, indexed by the two colors.
+
+  u32 cacheSet = cmprCacheSet(data.color1, data.color2);
+  u32 cacheTag = cmprCacheTag(data.color1, data.color2);
+
+  // our cache is initialized to all zeros, and we don't have space for a valid bit.
+  // This means tag 0 will end up as valid.
+  // A simple solution is to just never hit on tag zero
+
+  if (cmprCache[cacheSet].tag != cacheTag || cacheTag == 0)
+  {
+    TexDecoder_DecodePixelCMPR_Miss(data, cacheSet, cacheTag);
+  }
+
+  return cmprCache[cacheSet].colors;
+}
+
+static u32 TexDecoder_CMPRColorSelect(cmprTexelData data, u32* colors, int s, int t)
+{
+  int ss = s & 3;
+  int tt = t & 3;
+  int shift = tt * 8 + (6 - (ss << 1));
```

Comment:
This would be clearer as `((tt * 4) + (3 - ss)) * 2`, which hopefully would compile to the same thing.

Reply:
It produces different code. 

**clang x86_64:** Same number of instructions, but preforms 30% slower (according to MCA) due to dependencies.  
**gcc x86_64:** Same number of instructions, preforms the same according to MCA  
**clang arm64:** One extra instruction (MCA isn't working for arm on godbolt)
**gcc arm64:** One extra instruction (on top of GCC already producing a function that was one instruction longer)